yoyodyne/models/modules/lstm.py

```diff
@@ -0,0 +1,211 @@
+"""LSTM model classes."""
+
+from typing import Tuple
+
+import torch
+from torch import nn
+
+from ... import batches, defaults
+from . import attention, base
+
+
+class LSTMEncoder(base.BaseEncoder):
+    """Base encoder for LSTM."""
+
+    # Model arguments.
+    bidirectional: bool
+    # Constructed inside __init__.
+    module: nn.LSTM
+
+    def __init__(
+        self,
+        *args,
+        bidirectional=defaults.BIDIRECTIONAL,
+        **kwargs,
+    ):
+        """Initializes the encoder-decoder without attention.
+
+        Args:
+            *args: passed to superclass.
+            bidirectional (bool).
+            **kwargs: passed to superclass.
+        """
+        super().__init__(*args, **kwargs)
+        self.bidirectional = bidirectional
+        self.module = self.get_module()
+
+    def get_module(self):
+        return nn.LSTM(
+            self.embedding_size,
+            self.hidden_size,
+            num_layers=self.layers,
+            dropout=self.dropout,
+            batch_first=True,
+            bidirectional=self.bidirectional,
+        )
+
+    @property
+    def output_size(self):
+        return self.hidden_size * self.num_directions
+
+    @property
+    def num_directions(self):
+        return 2 if self.bidirectional else 1
+
+    def init_embeddings(
+        self, num_embeddings: int, embedding_size: int, pad_idx: int
+    ) -> nn.Embedding:
+        """Initializes the embedding layer.
+
+        Args:
+            num_embeddings (int): number of embeddings.
+            embedding_size (int): dimension of embeddings.
+            pad_idx (int): index of pad symbol.
+
+        Returns:
+            nn.Embedding: embedding layer.
+        """
+        return self._normal_embedding_initialization(
+            num_embeddings, embedding_size, pad_idx
+        )
+
+    def forward(
+        self, source: batches.PaddedTensor
+    ) -> Tuple[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]]:
+        """Encodes the input.
+
+        Args:
+            source (batches.PaddedTensor): source padded tensors and mask
+                for source, of shape B x seq_len x 1.
+
+        Returns:
+            Tuple[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]]:
+                encoded timesteps, and the LSTM h0 and c0 cells.
+        """
+        embedded = self.embed(source.padded)
+        # Packs embedded source symbols into a PackedSequence.
+        packed = nn.utils.rnn.pack_padded_sequence(
+            embedded, source.lengths(), batch_first=True, enforce_sorted=False
+        )
+        # -> B x seq_len x encoder_dim, (h0, c0).
+        packed_outs, (H, C) = self.module(packed)
+        encoded, _ = nn.utils.rnn.pad_packed_sequence(
+            packed_outs,
+            batch_first=True,
+            padding_value=self.pad_idx,
+            total_length=None,
+        )
+        return base.Output(encoded, hiddens=(H, C))
+
+
+class LSTMDecoder(LSTMEncoder):
```

Comment:
Inheriting from Encoder feels pretty odd to me. How much is shared? We *could* have both inherit from something like `LSTMModule` if there is substantial duplication. Or do you think that's unnecessary abstraction?

Reply:
Mmm, I think that should work. At least it would be more semantically coherrent.